### PR TITLE
[CGPO] Calibrated reward

### PR DESCRIPTION
**Pull request by gaetanlop**
# What does this PR do?

Adds a `get_calibrated_reward` function as introduced in the CGPO paper of Meta. Please refer to equation 5 in section 4.1.1 for more information on this (https://arxiv.org/pdf/2409.20370).

This PR should be part of a set of PRs to incorporate CGPO in trl.

Fixes # (issue)

#2156 

## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [X] Did you write any new necessary tests?

## Who can review?

@kashif @lewtun

### CODE REVIEW

**Comment by kashif** [comment]
the way i understood the calibrated reward was that the scores from a reward model might not be comparable across different completions given a prompt, and thus for some completion baseline or ground-truth completion `ā` for a prompt `s` and completion `a` the calibrated reward should be: `Rcalib(s, a) = σ(reward_model(s, a) − reward_model(s, ā))`

**Comment by gaetanlop** [comment]
Thanks for looking at it @kashif. Your code and mine are exactly the same except for the `missing_eos_penalty` part that I did not put in the function to be consistent with your function `get_reward` (we can handle the `missing_eos_penalty` part in the potential `CGPOTrainer` as done in the `OnlineDPOTrainer`). Apart from that, we have the same implementation. You are assuming that the `ground_truth_completion` do not contain the prompt, while I am assuming it contains it. 

Also, I am computing the reward for both `a` and `ā` in a single forward pass by concatenating them, while you are computing it separately for `a` and `ā`. 

Example using your naming conventions and assuming both `query_responses` and `baselines_responses` contain the prompt:
```python
batch_size = query_responses.shape[0]
concatenated_responses = torch.cat(
        (query_responses, baseline_responses),
        dim=0,
    )

reward_logits, final_rewards, sequence_lengths = get_reward(
    model, concatenated_responses, pad_token_id, context_length
)

generated_scores, ground_truth_scores = final_rewards.split(batch_size, dim=0)

final_rewards = F.sigmoid(generated_scores-ground_truth_scores)
```

For the returns, I am also returning all the `calibrated_logits`, and the `sequence_lengths` alongside what you are returning (final calibrated reward) to be consistent with the `get_reward` function of trl.



My implementation lacked a sigmoid function for the reward_logits thought. 

Am I correct?

**Comment by kashif** [comment]
ah right right! you are right!

**Comment by kashif** [comment]
so the reason i have the stuff split is because of padding... when i join the two different completions i have to pad them together while its slightly easier to pad each completion... and then i was scared if by contacting the memory needs might be too much for largish reward models... but yes makes sense

**Comment by gaetanlop** [comment]
Yes @kashif 100% agree, there are pros and cons for both methods. It also depends on the distributed training strategy you are using to train the model.

In any case, I checked the `trl` code base and it seems you adopted this concatenation method in the `OnlineDPOTrainer` https://github.com/huggingface/trl/blob/78249d9de46486a7fdb99c441ce0f52b9b0e1980/trl/trainer/online_dpo_trainer.py#L416-L427 

I think we should keep it this way. Wdyt?

**Comment by gaetanlop** [comment]
Closing in favor of #2190
